### PR TITLE
replace calls to libuv stop with faster deferred operations

### PR DESCRIPTION
**Pull request by vtjnash**
This avoids a Libuv bug in the Win32 handling of Polling on Sockets when a outdated firewall is installed. Plus, it should be faster.

~~TODO: A few of the tests for Pipe close likely need to be updated.~~
~~TODO: I forgot that the old setup for stop_reading was required in order to work around several NT kernel bugs. I'll keep it low performance on Windows and just make it faster on Unix (thus widening the gap between them even further).~~

### CODE REVIEW

File: base/deprecated.jl [hunk 1]
```diff
@@ -1076,4 +1076,9 @@ eval(Base.LinAlg, quote
     end
 end)
 
+
+@deprecate stop_reading(stream::LibuvStream) pause_reading(stream)
```
Comment (JeffBezanson) [inline base/deprecated.jl]: Using `@deprecate` will cause these to be exported, and they weren't before.

**Comment by tkelman** [comment]
can't backport new deprecations either

File: base/poll.jl [hunk 2]
```diff
@@ -257,78 +307,88 @@ end
...
+                return events
+            end
+        else
+            throw(events)
+        end
+        if fdw.refcount == (0, 0) # !open
+            events = EOFError()
+        else
+            start_watching(fdw) # make sure the poll is active
+            events = wait(fdw.notify)
```
Comment (tkelman) [inline base/poll.jl]: is the `wait` call already not type stable?

Comment (vtjnash) [inline base/poll.jl]: doesn't matter

File: base/deprecated.jl (hunk 1 above)
Comment (vtjnash) [inline base/deprecated.jl]: Upon reflection, renaming this is unnecessary. I can just slightly change how it is implemented to improve performance on Unix.

File: base/poll.jl (hunk 2 above)
Comment (tkelman) [inline base/poll.jl]: why not?

File: base/deprecated.jl [hunk 3]
```diff
@@ -1076,4 +1077,6 @@ eval(Base.LinAlg, quote
     end
 end)
 
+Filesystem.stop_watching(stream::Filesystem._FDWatcher) = depwarn("stop_watching(::_FDWatcher) is removed")
```
Comment (tkelman) [inline base/deprecated.jl]: This is not a useful deprecation warning. What should it be replaced with?

**Review by tkelman — DISMISSED** [review]

File: base/deprecated.jl (hunk 3 above)
Comment (vtjnash) [inline base/deprecated.jl]: It should be removed from the code. I highly doubt anyone was using it though, since it was also incorrect to be calling it before this removed it.

Comment (tkelman) [inline base/deprecated.jl]: ZMQ was using `stop_watching`, though possibly not on this type. This is still not a useful warning, you haven't addressed that.

**Review by tkelman — DISMISSED** [review]
deprecation warning message not useful and incorrectly calling `depwarn`. do not dismiss review without addressing it.

File: base/deprecated.jl (hunk 3 above)
Comment (tkelman) [inline base/deprecated.jl]: you're also not calling `depwarn` correctly here, it needs a second argument

Comment (vtjnash) [inline base/deprecated.jl]: nobody will see this, since nobody was using it, so it doesn't really matter what it says. but as it says, the correct action is to remove the call from the code.

The `stop_watching` in ZMQ is a different function.

Comment (vtjnash) [inline base/deprecated.jl]: should be good now:
```
julia> g = Base.Filesystem._FDWatcher(Libc.RawFD(0), true, false)
Base.Filesystem._FDWatcher(Ptr{Void} @0x00007fb46ab296d0,1,(2,0),Condition(Any[]),(false,false),0)

julia> Base.Filesystem.stop_watching(g)
WARNING: stop_watching(::_FDWatcher) is removed
 in depwarn(::String, ::Symbol) at ./deprecated.jl:64
 in stop_watching(::Base.Filesystem._FDWatcher) at ./deprecated.jl:1080
 in eval(::Module, ::Any) at ./boot.jl:236
 in eval(::Module, ::Any) at /Users/jameson/julia-master/usr/lib/julia/sys.dylib:?
 in eval_user_input(::Any, ::Base.REPL.REPLBackend) at ./REPL.jl:66
 in macro expansion at ./REPL.jl:97 [inlined]
 in (::Base.REPL.##3#4{Base.REPL.REPLBackend})() at ./event.jl:66
while loading no file, in expression starting on line 0
```

**Comment by tkelman** [comment]
the deprecation warning is still not useful. provide more information than "is removed."

and please answer the question about wait being type unstable.

**Comment by tkelman** [comment]
Please stop dismissing reviews, that should be done only when a reviewer is not actively responding to revisions. it was added so projects that have set pr's to require review for merging don't get blocked. reviewers should be the ones to modify their own review status.

**Comment by vtjnash** [comment]
The question about `wait` is nonsensical. Can you rephrase it into a meaningful question? Are you trying to ask whether `wait` must always return a particular value (it doesn't) or whether this point in the code base should expect a particular value (not particularly useful) or something else?

The deprecation message tells the zero users of this function that they should delete it from their (nonexistent) code. Can we please stop wasting time worrying about what this hypothetical user of a private Base API for a private base Type will do when an implementation detail changed and a some function become unnecessary?

**Comment by tkelman** [comment]
The code is written in a way that assigns the return value of wait to a variable, and branches based on the type of its result. Is it meant to be used that way? Is the type instability a problem, should we consider a different design in the future?

Yes this is minor, but if you're going to bother writing a warning, it should at least be a useful one. Something like "is not necessary and should be removed" would be better than 2 words.

**Comment by vtjnash** [comment]
> The code is written in a way that assigns the return value of wait to a variable, and branches based on the type of its result. Is it meant to be used that way?

yes

> Is the type instability a problem, should we consider a different design in the future?

no

> Yes this is minor, but if you're going to bother writing a warning, it should at least be a useful one. Something like "is not necessary and should be removed" would be better than 2 words.

"is removed" is concise and to the point, while "should be removed" is the same statement but with the wrong tense, and "is not necessary" is not correct. The most straightforward way to make this sentence longer is "... and should not be called", which is redundant with the definition of the word "removed". The even longer sentence would be "... and the code should be rewritten to use the documented API rather than depending on an implementation detail".

**Comment by tkelman** [comment]
if the warning is for people who might have been using it and is supposed to tell them what to do, a few more words of clarification would be better at accomplishing that. the tense on "is removed" doesn't make much sense either. has been removed from base, should be (but hasn't yet) from user code if they're seeing this warning. "is deprecated" works since deprecated is being used as an adjective, "is removed" doesn't work the same way.

**Comment by vtjnash** [comment]
hm, I'm not quite sure what to do with that failing test. It's not strictly correct (as can be seen by the platform-specifics of it), but is really just testing for an implementation detail in the kernel. So we could either (a) delete it (b) test it only on platforms where we have an expectation of what the kernel will do (c) compare against the result from a third party implementation (e.g. the python example below)

We can see how different kernels behave using a different language (python) for comparison:
```python
vtjnash@anubis:/data/vtjnash/julia$ uname -a
Linux anubis 4.8.0-26-generic #28-Ubuntu SMP Tue Oct 18 14:39:52 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
vtjnash@anubis:/data/vtjnash/julia$ python
Python 2.7.12+ (default, Sep 17 2016, 12:08:02) 
[GCC 6.2.0 20160914] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> x = os.pipe()
>>> x
(3, 4)
>>> import select
>>> select.select(x, x, x)
([], [4], [])
>>> 
```

```python
~/julia$ uname -a
Darwin Jamesons-MBP-Wifi.lan 16.1.0 Darwin Kernel Version 16.1.0: Thu Oct 13 21:26:57 PDT 2016; root:xnu-3789.21.3~60/RELEASE_X86_64 x86_64
...
([], [3, 4], [])
```

**Comment by tkelman** [comment]
Is the result something that we expect to be nondeterministic or vary depending on specific versions/distros/etc of Linux? We do shell out to perl in some tests on unix, but python isn't a test-time dependency right now.

Can we test for different expected results on different platforms?

File: base/deprecated.jl [hunk 4]
```diff
@@ -1086,4 +1087,6 @@ eval(Base.LinAlg, quote
     end
 end)
 
+Filesystem.stop_watching(stream::Filesystem._FDWatcher) = depwarn("stop_watching(::_FDWatcher) is removed", :stop_watching)
```
Comment (vtjnash) [inline base/deprecated.jl]: it's not just deprecated, it's removed (replaced by a warning that hopes the theoretical, non-existant caller wasn't expecting a side effect)

Comment (tkelman) [inline base/deprecated.jl]: It's in deprecated.jl, you're calling depwarn - it's a deprecation, just with a no-op replacement other than the warning. If it were removed it wouldn't be here at all.
